keps/sig-network/20191104-iptables-no-cluster-cidr.md

```diff
@@ -0,0 +1,282 @@
+
+### ipvs - accepting traffic after first packet, after being accepted by kubernetes rules
+
+The [rule here currently](https://github.com/kubernetes/kubernetes/blob/master/pkg/proxy/ipvs/proxier.go#L1635-L1654) looks as follows
+
+```go
+// The following two rules ensure the traffic after the initial packet
+// accepted by the "kubernetes forwarding rules" rule above will be
+// accepted, to be as specific as possible the traffic must be sourced
+// or destined to the clusterCIDR (to/from a pod).
+writeLine(proxier.filterRules,
+	"-A", string(KubeForwardChain),
+	"-s", proxier.clusterCIDR,
+	"-m", "comment", "--comment", `"kubernetes forwarding conntrack pod source rule"`,
+	"-m", "conntrack",
+	"--ctstate", "RELATED,ESTABLISHED",
+	"-j", "ACCEPT",
+)
+writeLine(proxier.filterRules,
+	"-A", string(KubeForwardChain),
+	"-m", "comment", "--comment", `"kubernetes forwarding conntrack pod destination rule"`,
+	"-d", proxier.clusterCIDR,
+	"-m", "conntrack",
+	"--ctstate", "RELATED,ESTABLISHED",
+	"-j", "ACCEPT",
+)
+```
+Again, applying similar logic to the last rule for iptables, the proposal here is to simplify this drop reference to the proxy.clusterCIDR and just match on the connection state.
+
+### Risks and Mitigations
+
+The biggest risk we have is that we are expanding the scope of the last rule to potentially include non-kubernetes traffic. This is considered mostly safe as it does not break any of the intended drop behavior. Plus once the initial connection has been accepted, assuming nodes are used for kubernetes workloads, it's highly unlikely that we would need to not accept it later.
+
+## Design Details
+
+The idea of ‘determine cluster originated traffic’ would be captured in a new interface type within kube-proxy, with different implementations of the interface. The kube-proxy implementation itself would just call method on interface to get the match criteria to write in the rule.
+
+This assumes that the match can be represented in a single rule. We want to avoid going down the path, as much as possible, of adding multiple rules. Support for multiple rules will be taken as a separate enhancement based on feedback as the current clusterCIDR rule is not yet being removed.
+
+### Graduation Criteria
+
+TODO - this will be resolved once the initial KEP has been reviewed. The current thinking is that these would just be additional flags to kube-proxy.
+
+## Implementation History
+
+2019-11-04 - Creation of the KEP
+
+## Drawbacks [optional]
+
+The main caveat in this KEP is the relaxation of the accept rule for "ESTABLISHED,RELATED" packets. The other two rules have equivalent implementations, as long as we continue to guarantee that pod traffic is routed at the node boundary on _every_ and _all_ nodes that makes up the kubernetes cluster. This would not work if that assumption were to change.
+
+## Alternatives [optional]
```

Comment:
I agree with your initial assessment, @satyasm, that ClusterCIDR is, functionally, equivalent to setting NonMasqueradeCIRDS, and that all other uses of it can be refactored away. So what if we just run with that?

What if we added a `NonMasqueradeCIDRs[]` field to the KubeProxyConfig object? What if we added it to the Node object? Then, network plugins / operators could configure that at will. For networks that are routed to the outside world, the only entry would be `0.0.0.0/0`. For more interesting topologies, it could be customized.

Thoughts, @danwinship?

Reply:
Another name could be "ReachableCIDRs" or "RoutableCIDRs" to indicate that these are the network blocks for which the pod + service networks are directly reachable.

Reply:
Right. Though the thinking is that if we can change the rules to not have to track the cluster CIDR at all, then it will give us much more flexibility on doing IPAM. For example, say as a next step we work to standardize representation of multiple node.podCIDR for an address family. In that case, if we have N CIDRs per node and M CIDRs for the cluster, it becomes a NxM combination to track. While if we move to non-ip-based detection, we don't have to worry about the NxM combination (simplifies rules?).

In addition, based on the all the discussion at kubecon around 5G etc and additional network interfaces, I think it makes even more sense to pivot to interface name based to separate what is kubernetes control plane network interface v/s other use cases?

Reply:
added this as "ip-masq-agent like behavior" under alternatives.

Reply:
(hey, back from the holidays, whee)
I guess my feeling is that using something like interface name(s) is an unnecessary coupling between layers of the stack. Interface name is a layer-1 concept, whereas IP blocks is really in the domain of layer 3.

Now, if we wanted to somehow autodetect by correctly programming the kernel routing table, that's something that makes sense to me. Because that's how we communicate layer-3 information to the kernel.

Reply:
agreed that it does feel wrong to mix layers if we consider them as independent variables. But as noted above, we do orchestrate the setup and can _influence_ the naming convention for these interfaces. that being the case, it's not unreasonable to assume that we can set the names up as needed as part of the contract for initialization? From that perspective, it might not be a bad idea to go by interface names and not depend on ip addresses, thus providing more independence for downstream operations related to IPAM?

Reply:
I'm not sure I understand the discussion of layers here - we're trying to define a pod's node locality, right? I don't think that implies layer 3 in any way, unless I'm misunderstanding something?

Reply:
I understood it as "if we are talking about ip routability", why are we trying to co-ordinate something at a lower level to do so. @caseydavenport I'm approaching it from your perspective, which is "detect it was local traffic" to influence routing decisions. And from that perspective too, how we detect it is local need not be ip based I guess, though the decision (impact) is in terms of IP.

Reply:
@satyasm yeah, that's the kind of layer-mixing I was thinking of, thanks.

If everyone thinks I'm totally off the mark, I'll accept that. But something seems a bit off to me.

Reply:
Can this be marked as resolved given overall alignment so far?

Reply:
I don't think this is a problem - it's offering the option to cross layers in order to simplify.  Users can't really mix-and-match parts of their net stack without thinking about it anyway.  This gives thoughtful installs a better option.